gokart/gcs_obj_metadata_client.py

```diff
@@ -71,30 +81,63 @@ def add_task_state_labels(
             if update_response is None:
                 logger.error(f'failed to patch object {obj} in bucket {bucket} and object {obj}.')
 
+    @staticmethod
+    def _normalize_labels(labels: Optional[dict[str, str]]) -> dict[str, str]:
+        def _normalize_labels_helper(_params: Optional[dict[str, Any]]) -> dict[str, str]:
+            return {str(key): str(value) for key, value in labels.items()} if labels else {}
+
+        return _normalize_labels_helper(labels)
+
     @staticmethod
     def _get_patched_obj_metadata(
         metadata: Any,
-        task_params: Optional[dict[Any, str]] = None,
+        task_params: Optional[dict[str, str]] = None,
+        custom_labels: Optional[dict[str, Any]] = None,
     ) -> Union[dict, Any]:
         # If metadata from response when getting bucket and object information is not dictionary,
         # something wrong might be happened, so return original metadata, no patched.
         if not isinstance(metadata, dict):
             logger.warning(f'metadata is not a dict: {metadata}, something wrong was happened when getting response when get bucket and object information.')
             return metadata
 
-        if not task_params:
+        if not task_params and not custom_labels:
             return metadata
         # Maximum size of metadata for each object is 8 KiB.
         # [Link]: https://cloud.google.com/storage/quotas#objects
-        max_gcs_metadata_size, total_metadata_size, labels = 8 * 1024, 0, []
-        for label_name, label_value in task_params.items():
+        max_gcs_metadata_size, total_metadata_size = 8 * 1024, 0
+        labels: list[tuple[str, str]] = []
+        has_seen_keys: set[str] = set()
+
+        normalized_task_params_labels = GCSObjectMetadataClient._normalize_labels(task_params)
+        normalized_custom_labels = GCSObjectMetadataClient._normalize_labels(custom_labels)
+        # There is a possibility that the keys of user-provided labels(custom_labels) may conflict with those generated from task parameters (task_params_labels).
+        # However, users who utilize custom_labels are no longer expected to search using the labels generated from task parameters.
+        # Instead, users are expected to search using the labels they provided.
+        # Therefore, in the event of a key conflict, the value registered by the user-provided labels will take precedence.
+        total_metadata_size, labels = GCSObjectMetadataClient._add_labels_to_metadata(
+            normalized_custom_labels, total_metadata_size, max_gcs_metadata_size, labels, has_seen_keys
+        )
+        _, labels = GCSObjectMetadataClient._add_labels_to_metadata(
+            normalized_task_params_labels, total_metadata_size, max_gcs_metadata_size, labels, has_seen_keys
+        )
+        return dict(metadata) | dict(labels)
+
+    @staticmethod
+    def _add_labels_to_metadata(
```

Comment:
hmm, mutable updating `has_seen_keys` can made some bugs.
why don't you return `has_seen_keys` from `_add_labels_to_metadata` like following?
```suggestion
    def _add_labels_to_metadata(
            labels_dict: dict[str, str], total_metadata_size: int, max_gcs_metadata_size: int, labels: Optional[list[tuple[str, str]]] = None, Optional[has_seen_keys: set[str]] = None,
    ) -> tuple[int, list[tuple[str, str]]]:
        labels = copy.copy(labels) if labels else []
        has_seen_keys = copy.copy(has_seen_keys) if has_seen_keys else set({})
        ...
        return total_metadata_size, labels, has_seen_keys
```

Reply:
Ahh. As you said, should return has_seen_keys as same as labels.
But I have one question.
In python, set, list, dict, they are passed as reference.
But in your implementation, why do you suggest to copy it?
For immuability? I'm wondering what is the best practice in some kind of this case.

Reply:
@TlexCypher 
As you said, they are passed as reference, which also update original variables of arguments.
For example, the following assertion fails since the original `l` is also updated.

```python
def update_list(l: list[str]) -> list[str]:
    l.append('new')
    return l


l = ['old']
new_l = update_list(l)

assert len(l) < len(new_l)
```

This makes sometimes developer confused, e.g. comparing original list and new one.
The function name `_add_labels_to_metadata` does not imply to update passed arguments, but just return new 
metadata.
For many cases, I prefer make variable to be immutable, thus I suggested to copy them

Reply:
Thank you for great help.
I can clearly understand why immutability is necessary.
Thus, I accept your suggestion!